### ISSUE TITLE
tlstran_pipe_nego_cb: connect nego error rv: Cryptographic error

### ISSUE THREAD

**Issue by amcharhal** [nanomq/nanomq#1764]
Problem with TLS when attempting to connect many clients simultaneously.

SDK client: Node.js MQTT client 
Protocol: SSL with certs 

Reproduction: Attempting to connect multiple clients simultaneously using the TLS protocol and a certificate.

Result: The broker becomes blocked and unable to connect to any client. The only solution is to restart the server."

debug logs mode 
show only these lines at each attempt 
2024-04-23 07:32:00 [14481] WARN  /home/runner/work/nanomq/nanomq/nng/src/sp/transport/mqtts/broker_tls.c:310 tlstran_pipe_nego_cb: nego aio error: Cryptographic error
2024-04-23 07:32:00 [14481] ERROR /home/runner/work/nanomq/nanomq/nng/src/sp/transport/mqtts/broker_tls.c:448 tlstran_pipe_nego_cb: connect nego error rv: Cryptographic error(26)

**Comment by amcharhal** [comment]
is that naomq has maximum number of connection using TLS ?

**Comment by thomascenni** [comment]
> The broker becomes blocked and unable to connect to any client. The only solution is to restart the server."

Hello, I have exactly the same problem, even with recent versions of nanomq. I am connecting 3 or 4 clients manually, so it's more than 500ms delay. I am ausing the same certificate for all the clients, can this be related to the root cause ? 
This happens with the 0.22.x release, it doesn't happen with 0.21.x release.

**Comment by JaylinYu** [comment]
> > The broker becomes blocked and unable to connect to any client. The only solution is to restart the server."
> 
> 
> 
> Hello, I have exactly the same problem, even with recent versions of nanomq. I am connecting 3 or 4 clients manually, so it's more than 500ms delay. I am using the same certificate for all the clients, can this be related to the root cause ? 
> 
> This happens with the 0.22.x release, it doesn't happen with 0.21.x release.

Are you using vanila TCP/TLS or websocket?
If your issue is not happening before 0.22.then The newly synchronization with upstream is highly suspicious.

**Comment by thomascenni** [comment]
> Are you using vanila TCP/TLS or websocket? If your issue is not happening before 0.22.then The newly synchronization with upstream is highly suspicious.

I'm using TLS, I have disabled WS with NANOMQ_WEBSOCKET_ENABLE=false

**Comment by JaylinYu** [comment]
btw, i really don't recommend you compile from master branch directly, better pick a tag.

**Comment by edaspb** [comment]
Thank you @JaylinYu for the comments.
Few more details: SSL listeners stop responding to the legitimate clients after random amount of time, like few hours after service start. Tcpdump see the traffic go to and from port 8883.
No logs.
But if I try to make a connection yousing wrong certificate broker_tls.c generate a regular error.
TCP listener keeps working. SSL recuperates after nanomq restart.
I will try to recompile nanomq from last stable tag.

**Comment by JaylinYu** [comment]
If there is no traffic being captured， then it is absolute normal that no log coming from tls listener. And it not tls listener to be blamed as stop responding.

**Comment by edaspb** [comment]
@JaylinYu, the traffic does go to and from the listener...

**Comment by JaylinYu** [comment]
ok
Will try to reproduce it.

**Comment by edaspb** [comment]
I'm not sure if the next messages are related to the issue. What I have in the log:

```
2024-10-12 10:41:10 [10] DEBUG /nanomq/nng/src/sp/protocol/mqtt/mqtt_parser.c:612 conn_handler: remain len 55 max len 57 prop len 2 pos 12
2024-10-12 10:41:10 [10] DEBUG /nanomq/nng/src/supplemental/mqtt/mqtt_codec.c:4075 decode_buf_properties: remain len 55 prop len 2 curpos 0x7fd15000a74d endpos 0x7fd15000a74f
2024-10-12 10:41:26 [13] DEBUG /nanomq/nng/src/sp/protocol/mqtt/mqtt_parser.c:612 conn_handler: remain len 55 max len 57 prop len 2 pos 12
2024-10-12 10:41:26 [13] DEBUG /nanomq/nng/src/supplemental/mqtt/mqtt_codec.c:4075 decode_buf_properties: remain len 55 prop len 2 curpos 0x7fd15400688d endpos 0x7fd15400688f
2024-10-12 10:41:31 [16] DEBUG /nanomq/nanomq/apps/broker.c:338 server_cb: RECV  ^^^^ ctx7 ^^^^

2024-10-12 10:41:31 [16] DEBUG /nanomq/nanomq/apps/broker.c:635 server_cb: WAIT ^^^^ ctx7 ^^^^
2024-10-12 10:41:31 [16] DEBUG /nanomq/nanomq/apps/broker.c:722 server_cb: broker has nothing to do
2024-10-12 10:41:44 [15] DEBUG /nanomq/nng/src/sp/protocol/mqtt/mqtt_parser.c:612 conn_handler: remain len 55 max len 57 prop len 2 pos 12
2024-10-12 10:41:44 [15] DEBUG /nanomq/nng/src/supplemental/mqtt/mqtt_codec.c:4075 decode_buf_properties: remain len 55 prop len 2 curpos 0x7fd14400207d endpos 0x7fd14400207f
2024-10-12 10:41:56 [11] DEBUG /nanomq/nanomq/apps/broker.c:338 server_cb: RECV  ^^^^ ctx6 ^^^^

2024-10-12 10:41:56 [14] DEBUG /nanomq/nanomq/apps/broker.c:338 server_cb: RECV  ^^^^ ctx3 ^^^^

2024-10-12 10:41:56 [10] DEBUG /nanomq/nanomq/apps/broker.c:635 server_cb: WAIT ^^^^ ctx6 ^^^^
2024-10-12 10:41:56 [15] DEBUG /nanomq/nanomq/apps/broker.c:635 server_cb: WAIT ^^^^ ctx3 ^^^^
2024-10-12 10:41:56 [10] DEBUG /nanomq/nanomq/apps/broker.c:722 server_cb: broker has nothing to do
2024-10-12 10:41:56 [15] DEBUG /nanomq/nanomq/apps/broker.c:722 server_cb: broker has nothing to do
2024-10-12 10:41:59 [16] DEBUG /nanomq/nanomq/apps/broker.c:338 server_cb: RECV  ^^^^ ctx2 ^^^^

2024-10-12 10:41:59 [16] DEBUG /nanomq/nanomq/apps/broker.c:635 server_cb: WAIT ^^^^ ctx2 ^^^^
2024-10-12 10:41:59 [16] DEBUG /nanomq/nanomq/apps/broker.c:722 server_cb: broker has nothing to do
2024-10-12 10:42:15 [12] DEBUG /nanomq/nng/src/sp/protocol/mqtt/mqtt_parser.c:612 conn_handler: remain len 55 max len 57 prop len 2 pos 12
2024-10-12 10:42:15 [12] DEBUG /nanomq/nng/src/supplemental/mqtt/mqtt_codec.c:4075 decode_buf_properties: remain len 55 prop len 2 curpos 0x7fd13800670d endpos 0x7fd13800670f
2024-10-12 10:42:31 [15] DEBUG /nanomq/nanomq/apps/broker.c:338 server_cb: RECV  ^^^^ ctx8 ^^^^

2024-10-12 10:42:31 [15] DEBUG /nanomq/nanomq/apps/broker.c:635 server_cb: WAIT ^^^^ ctx8 ^^^^
2024-10-12 10:42:31 [15] DEBUG /nanomq/nanomq/apps/broker.c:722 server_cb: broker has nothing to do
```

**Comment by edaspb** [comment]
> ok Will try to reproduce it. Mbedtls is highly suspicious, mosq using openssl instead.
I have tried different version of mbedtls starting from Ubuntu 24.04 reps to the last self-compiled released one. No changes.
> Any signs before tls lisntener went into frozen mode? Seeing crypto error?
No, no logs with 'warn' mode so far.

**Comment by edaspb** [comment]
@JaylinYu, I would like to ask you temporary close the issue, I found a fiscal disk error on my cloud provider platform. I think it would be better first to solve it, highly probable that is the cause. Again, thanks a lot for your support, I will update the status further.

**Comment by edaspb** [comment]
No, issue with the hard drive wasn't the cause. BTW I could reproduce the describing behavior.
On my rollout the ssl listener stops responding to legit client requests  after first unsuccessful request using wrong certs. To reproduce:
1. Create distinct certificate chain. I was following official [instruction](https://nanomq.io/docs/en/latest/tutorial/SSL_TLS.html). 
2. Make a connection attempt to mqtts service using new not legit key/cert pair. The expected error will occur.
3. Make a request with proper certificate.

I have reproduced such behavior with self-compiled instance and with official docker container emqx/nanomq:0.22.8-full

**Comment by JaylinYu** [comment]
MBedtls is innocent; this was my mistake to wrongly cancel the listener if I received an illegitimate certificate.
Will push a fix and release 0.22.9

@edaspb  Thank you for the detailed reproduce steps, I will not be able to identify this bug without your help.
TLS debugging is really misleading, I was always take this as an normal crypto error. Didn't expect the afterward mistake.

**Comment by JaylinYu** [comment]
the fix is also wrapped into 0.22.8
